### PR TITLE
feat(application): add delete feature

### LINKED ISSUE #32: Add delete application feature to dashboard
image

add a red color delete or remove icon on the right side of the application card so as to remove or delete that application.

image

### PR DESCRIPTION
**Pull request by MASTERCHIEF011**
Fixes #32

### CODE REVIEW

**Comment by YashKumarVerma** [comment]
> Fixes #32

add changes done in this pr

File: src/application/application.service.ts [hunk 1]
```diff
@@ -33,6 +33,10 @@ export class ApplicationService {
     }
   }
 
+  async deleteApplication(id: String) {
```
Comment (YashKumarVerma) [inline src/application/application.service.ts:36]: no need to name it as "deleteApplication" as it is the service for application itself and it is being redundant.

File: src/dashboard/dashboard.controller.ts [hunk 2]
```diff
@@ -75,4 +75,10 @@ export class DashboardController {
       },
     });
   }
+
+  @Post('/dev/:id')
+  async deleteApp(@Res() res: Response, @Param('id') id: String) {
+    const action = await this.applicationService.deleteApplication(id);
+    res.redirect('/dashboard/dev');
+  }
```
Comment (YashKumarVerma) [inline src/dashboard/dashboard.controller.ts:83]: add validation to check if admin is sending a request to delete.

**Review by YashKumarVerma — CHANGES_REQUESTED** [review]

File: src/dashboard/dashboard.controller.ts (hunk 2 above)
Comment (MASTERCHIEF011) [reply src/dashboard/dashboard.controller.ts:83]: okk

File: src/application/application.service.ts [hunk 3]
```diff
@@ -33,6 +33,18 @@ export class ApplicationService {
...
+      const user = await this.applicationModel.findById({ _id: id });
+      if (JSON.stringify(user.admin) === JSON.stringify(authorizedUser.id)) {
+        const deleteApp = await this.applicationModel.findByIdAndDelete({ _id: id });
+      }
+    } catch (e) {
+      this.logger.error(e);
+      throw new ConflictException(e.message);
+    }
+  }
+
```
Comment (YashKumarVerma) [inline src/application/application.service.ts:47]: the logic to test whether user is authorized or not to delete an application must reside in the controller

**Review by YashKumarVerma — CHANGES_REQUESTED** [review]

File: src/application/application.service.ts (hunk 3 above)
Comment (MASTERCHIEF011) [reply src/application/application.service.ts:47]: okay

File: src/application/application.service.ts [hunk 4]
```diff
@@ -33,6 +33,15 @@ export class ApplicationService {
     }
   }
 
+  async delete(id: String) {
```
Comment (YashKumarVerma) [inline src/application/application.service.ts:36]: the id is of type ObjectId

File: src/dashboard/dashboard.controller.ts [hunk 5]
```diff
@@ -75,4 +75,15 @@ export class DashboardController {
       },
     });
   }
+
+  @Post('/dev/:id')
+  @UseGuards(JwtAuthGuard)
+  async delete(@Request() req, @Res() res: Response, @Param('id') id: string) {
+    const loggedInUser: LoggedInUser = req.user;
+    const user = await this.applicationService.findOneById(id);
+    if (JSON.stringify(user.admin) === JSON.stringify(loggedInUser.id)) {
```
Comment (YashKumarVerma) [inline src/dashboard/dashboard.controller.ts:84]: could you explain what is the meaning of 
```js
if (JSON.stringify(user.admin) === JSON.stringify(loggedInUser.id)) {
```

**Review by YashKumarVerma — COMMENTED** [review]
Rest LGTM

File: src/dashboard/dashboard.controller.ts (hunk 5 above)
Comment (MASTERCHIEF011) [reply src/dashboard/dashboard.controller.ts:84]: one was of type objectId and the other was of type String so to make their types equal i converted them into a Json String.

Comment (MASTERCHIEF011) [reply src/dashboard/dashboard.controller.ts:84]: Here I am checking whether the current user i.e (loggedInUser.id) who is giving the delete command is same as the one who created it i.e (user.admin).

Comment (YashKumarVerma) [reply src/dashboard/dashboard.controller.ts:84]: could you check the datatype of `user.admin` ?

Comment (MASTERCHIEF011) [reply src/dashboard/dashboard.controller.ts:84]: Upon executing the below command 
image

I get this as a result:
image
So i guess that user.admin is of object type.

**Review by YashKumarVerma — APPROVED** [review]

**Review by YashKumarVerma — APPROVED** [review]

**Comment by YashKumarVerma** [comment]
thanks !

**Pull request merged**